cpp/ql/src/Critical/ScanfChecks.qll

```diff
@@ -0,0 +1,35 @@
+private import cpp
+private import semmle.code.cpp.commons.Scanf
+
+private predicate exprInBooleanContext(Expr e) {
+  e.getParent() instanceof BinaryLogicalOperation
+  or
+  e.getParent() instanceof UnaryLogicalOperation
+  or
+  e = any(IfStmt ifStmt).getCondition()
+  or
+  e = any(WhileStmt whileStmt).getCondition()
+  or
+  exists(EqualityOperation eqOp, Expr other |
+    eqOp.hasOperands(e, other) and
+    other.getValue() = "0"
+  )
+  or
+  exists(Variable v |
+    v.getAnAssignedValue() = e and
+    forex(Expr use | use = v.getAnAccess() | exprInBooleanContext(use))
+  )
+}
```

Comment:
This code doesn't quite work as with `if(x)` the IR doesn't compare `x` against zero as that is folded into the `ConditionalBranchInstruction` itself. Just doing IR pattern matching gets further but there are still more missing cases.

The additional cases all seem to be value numbering matches which don't matter much here.

Reply:
Hm. What is the testcase that would fail with my suggestion? Is it something along the lines of:
```cpp
int i = 0;
int r = scanf("%d", &i);
if (r) {
  use(i);
}
```
?

In that case, would it be a sufficient to change the suggestion to be:
```ql
private predicate exprInBooleanContext(Expr e) {
  exists(IRGuardCondition gc |
    exists(Instruction i, ConstantInstruction zero |
      zero.getValue() = "0" and
      i.getUnconvertedResultExpression() = e and
      gc.comparesEq(valueNumber(i).getAUse(), zero.getAUse(), 0, _, _)
    )
    or
    gc.getUnconvertedResultExpression() = e
  )
}
```
?